src/main/java/net/snowflake/ingest/streaming/internal/SubscopedTokenExternalVolumeManager.java

```diff
@@ -0,0 +1,230 @@
+    }
+    logger.logDebug(
+        "Created SubscopedTokenExternalVolumeManager with clientName=%s and clientPrefix=%s",
+        clientName, clientPrefix);
+  }
+
+  /**
+   * Given a fully qualified table name, return the target storage by looking up the table name
+   *
+   * @param fullyQualifiedTableName the target fully qualified table name
+   * @return target storage
+   */
+  @Override
+  public InternalStage getStorage(String fullyQualifiedTableName) {
+    // Only one chunk per blob in Iceberg mode.
+    return getVolumeSafe(fullyQualifiedTableName);
+  }
+
+  /** Informs the storage manager about a new table that's being ingested into by the client. */
+  @Override
+  public void registerTable(TableRef tableRef) {
+    if (this.externalVolumeMap.containsKey(tableRef.fullyQualifiedName)) {
+      logger.logInfo(
+          "Skip registering table since its already been registered with the VolumeManager."
+              + " tableRef=%s",
+          tableRef);
+      return;
+    }
+
+    // future enhancement - per table locks instead of per-client lock
+    synchronized (registerTableLock) {
+      if (this.externalVolumeMap.containsKey(tableRef.fullyQualifiedName)) {
+        logger.logInfo(
+            "Skip registering table since its already been registered with the VolumeManager."
+                + " tableRef=%s",
+            tableRef);
+        return;
+      }
+
+      // get the locationInfo when we know this is the first register call for a given table.
+      // This is done to reduce the unnecessary overload on token generation if a client opens up a
+      // hundred channels at
+      // the same time.
+      FileLocationInfo locationInfo = getRefreshedLocation(tableRef, Optional.empty());
+
+      try {
+        InternalStage externalVolume =
+            new InternalStage(
+                this,
+                clientName,
+                getClientPrefix(),
+                tableRef,
+                locationInfo,
+                DEFAULT_MAX_UPLOAD_RETRIES);
+        this.externalVolumeMap.put(tableRef.fullyQualifiedName, externalVolume);
+      } catch (SFException ex) {
+        logger.logError(
+            "ExtVolManager.registerTable for tableRef=% failed with exception=%s", tableRef, ex);
+        // allow external volume ctor's SFExceptions to bubble up directly
+        throw ex;
+      } catch (Exception err) {
+        logger.logError(
+            "ExtVolManager.registerTable for tableRef=% failed with exception=%s", tableRef, err);
+        throw new SFException(
+            err,
+            ErrorCode.UNABLE_TO_CONNECT_TO_STAGE,
+            String.format("fullyQualifiedTableName=%s", tableRef));
+      }
+    }
+  }
+
+  @Override
+  public BlobPath generateBlobPath(String fullyQualifiedTableName) {
+    InternalStage volume = getVolumeSafe(fullyQualifiedTableName);
+
+    // {nullableTableBasePath}/data/streaming_ingest/{figsId}/snow_{volumeHash}_{figsId}_{workerRank}_1_
```

Comment:
In the non iceberg path I recently added a check to make sure that the deployment id returned when refreshing the location matches the one that the channel was originally created with. This is to prevent a channel from ingesting data into the wrong deployment in replication scenarios. Do we need something like that for iceberg?

Reply:
Yes, already handling that with iceberg. THe workerRank is populated with the deployment id by the service, and is verified in registerBlob() for being correct.So a filepath sent down by deployment 1 cannot be used to register a file in deployment 2.

btw this depl id fix was only needed when doing client side encryption from what i understand, still added it in here for consistent behaviors across iceberg and non-iceberg.

Reply:
Thanks. It was not just for encryption - it also prevents some race conditions where blobs are uploaded in one deployment but registered in another and hence gets lost.